backend/optimize.ml

```diff
@@ -0,0 +1,210 @@
+  let random_value = Random.int (1 lsl 24) in
+  Printf.sprintf "%06x" random_value
+
+let set_csv () =
+  if Option.is_none !csv_singleton
+  then (
+    let new_csv = IntCsv.create ["remove_useless_mov"; "fold_intop_imm"] in
+    if !Flambda_backend_flags.cfg_peephole_optimize_track
+    then Stdlib.at_exit (fun () -> 
+      Random.self_init (); 
+      (* Generate a random 6-hexdigit string for the name of the csv file *)
+      IntCsv.print new_csv (random_hex_string () ^ ".csv"));
+    csv_singleton := Some new_csv)
+
+let get_csv () = Option.get !csv_singleton
+
+(* CR gtulba-lecu for gtulba-lecu: make sure that this comparison *)
+let are_equal_regs reg1 reg2 =
+  Reg.same_loc reg1 reg2 && Regalloc_utils.same_reg_class reg1 reg2
+
+let update_csv str =
+  let csv = get_csv () in
+  match IntCsv.rows csv with
+  | [] ->
+    (* there should always be at least a row in the csv when updating *)
+    assert false
+  | rows_hd :: rows_tl ->
+    IntCsv.set_rows csv (IntCsv.update_row rows_hd str IntCell.update :: rows_tl)
+
+(* Logical condition for simplifying the following case: 
+  {v 
+    mov x, y 
+    mov y, x 
+  v}
+
+   In this case, the second instruction should be removed *)
+
+let remove_useless_mov (fst : Cfg.basic Cfg.instruction DLL.cell)
+    (snd : Cfg.basic Cfg.instruction DLL.cell) =
+  let fst_val = DLL.value fst in
+  let snd_val = DLL.value snd in
+  match fst_val.desc with
+  | Op Move | Op Spill | Op Reload -> (
+    let fst_src, fst_dst = fst_val.arg.(0), fst_val.res.(0) in
+    match snd_val.desc with
+    | Op Move | Op Spill | Op Reload ->
+      let snd_src, snd_dst = snd_val.arg.(0), snd_val.res.(0) in
+      if are_equal_regs fst_src snd_dst && are_equal_regs fst_dst snd_src
+      then (
+        if !Flambda_backend_flags.cfg_peephole_optimize_track
+        then update_csv "remove_useless_mov";
+        DLL.delete_curr snd;
+        Some fst)
+      else None
+    | _ -> None)
+  | _ -> None
+
+(* Logical condition for simplifying the following case: 
+  {v 
+    <op 1> const1, r 
+    <op 2> const2, r 
+  v}
+
+  to: 
+  {v 
+    <op 1> (const1 <op 2> const2), r 
+  v}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+(* CR-soon gtulba-lecu for gtulba-lecu: implement the rest of Intop_imm cases 
+   (only Imulh left to optimize if possible)
+    
+    it may be the case that after Iadd, Isub folding the result is 0, treat this case.
+*)
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1: Mach.integer_operation), (op2: Mach.integer_operation) with
+  | Mach.Iand, Mach.Iand -> Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior -> Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor -> Some (Mach.Ixor, imm1 lxor imm2)
+  | Ilsl, Ilsl -> Some (Mach.Ilsl, min (imm1 + imm2) Sys.int_size)
+  | Ilsr, Ilsr -> Some (Mach.Ilsr, min (imm1 + imm2) Sys.int_size)
+  | Iasr, Iasr -> Some (Mach.Iasr, min (imm1 + imm2) Sys.int_size)
```

Comment:
I am not positive it is correct; if memory
serves `Sys.int_size` will be 63, so that
would for instance fold two `lsl 32` to
`lsl 63`.